File: src/core/landmarks.py

"""
Single Source of Truth para landmarks cefalométricos.

Define nombres, índices y funciones auxiliares para análisis cefalométrico.
Índices basados en el estándar del dataset Aariz (29 landmarks).
"""

from typing import Dict, Tuple, Optional

# Índices de landmarks (0-28) según estándar Aariz
LANDMARK_NAMES: Tuple[str, ...] = (
    "A", "ANS", "B", "Me", "N", "Or", "Pog",
    "PNS", "Pn", "R", "S", "Ar", "Co", "Gn",
    "Go", "Po", "LPM", "LIT", "LMT", "UPM",
    "UIA", "UIT", "UMT", "LIA", "Li", "Ls",
    "N'", "Pog'", "Sn"
)

NUM_LANDMARKS = len(LANDMARK_NAMES)

# Índices clave para análisis Steiner/Ricketts
IDX_SELLA = 10      # S
IDX_NASION = 4      # N
IDX_A_POINT = 0     # A
IDX_B_POINT = 2     # B
IDX_POGONION = 6    # Pog

# Mapeo nombre -> índice
NAME_TO_IDX: Dict[str, int] = {name: i for i, name in enumerate(LANDMARK_NAMES)}

# Nombres completos (opcional, para reportes)
FULL_NAMES: Dict[str, str] = {
    "A": "Punto A (Subespinale)",
    "ANS": "Anterior Nasal Superior",
    "B": "Punto B (Supramental)",
    "Me": "Mentón (Menton)",
    "N": "Nasion",
    "Or": "Orbitale",
    "Pog": "Pogonion",
    "PNS": "Posterior Nasal Spine",
    "Pn": "Prosthion",
    "R": "Ridge",
    "S": "Sella Turcica",
    "Ar": "Articulación",
    "Co": "Cóndilo",
    "Gn": "Gnathion",
    "Go": "Gonion",
    "Po": "Porion",
    "LPM": "Low Point Mandible",
    "LIT": "Lower Incisor Tip",
    "LMT": "Lower Molar Tip",
    "UPM": "Upper Molar Tip",
    "UIA": "Upper Incisor Apex",
    "UIT": "Upper Incisor Tip",
    "UMT": "Upper Molar Tip",
    "LIA": "Lower Incisor Apex",
    "Li": "Lower Incisor",
    "Ls": "Lips",
    "N'": "N Prime",
    "Pog'": "Pogonion Prime",
    "Sn": "Subnasale"
}


def get_index(name: str) -> Optional[int]:
    """
    Obtiene el índice de un landmark por nombre.

    Args:
        name: Nombre del landmark (ej. 'S', 'N', 'A', 'B', 'Go')

    Returns:
        Índice (0-28) o None si no existe.
    """
    return NAME_TO_IDX.get(name)


def get_name(index: int) -> str:
    """
    Obtiene el nombre de un landmark por índice.

    Args:
        index: Índice del landmark (0-28)

    Returns:
        Nombre del landmark o 'LM{index}' si fuera de rango.
    """
    if 0 <= index < NUM_LANDMARKS:
        return LANDMARK_NAMES[index]
    return f"LM{index}"


def validate_indices(indices):
    """
    Valida que todos los índices estén en rango [0, NUM_LANDMARKS).

    Args:
        indices: Lista o tupla de índices

    Raises:
        ValueError: Si algún índice es inválido.
    """
    for idx in indices:
        if not (0 <= idx < NUM_LANDMARKS):
            raise ValueError(
                f"Índice {idx} fuera de rango. "
                f"Debe estar en [0, {NUM_LANDMARKS-1}]"
            )


# Tuplas nombradas para ángulos comunes
LANDMARK_ANGLE_SNA = (IDX_SELLA, IDX_NASION, IDX_A_POINT)  # S-N-A
LANDMARK_ANGLE_SNB = (IDX_SELLA, IDX_NASION, IDX_B_POINT)  # S-N-B
LANDMARK_ANGLE_POG = (IDX_SELLA, IDX_NASION, IDX_POGONION)  # S-N-Pog


def angle_key_points(key1: str, key2: str, key3: str) -> Tuple[int, int, int]:
    """
    Devuelve los índices para tres puntos clave dados por nombre.

    Args:
        key1, key2, key3: Nombres de landmarks

    Returns:
        Tupla (idx1, idx2, idx3)

    Raises:
        KeyError: Si algún nombre no existe.
    """
    return (NAME_TO_IDX[key1], NAME_TO_IDX[key2], NAME_TO_IDX[key3])

File: frontend/src/constants/landmarks.js
// 29 landmarks - ORDEN EXACTO según src/core/landmarks.py (Aariz dataset)
// CRÍTICO: El modelo PyTorch usa ESTE orden. Cualquier desviación causa mapeo incorrecto.
// Categorías: cranial (red #EF4444), mandibular (blue #3B82F6), dental (green #22C55E), soft-tissue (yellow #EAB308)

const LANDMARKS = [
  // 0: A (Subespinale) - Cranial
  { id: 0, name: 'A-point (Subespinale)', category: 'cranial', color: '#EF4444' },
  // 1: ANS (Anterior Nasal Spine) - Cranial
  { id: 1, name: 'ANS (Anterior Nasal Spine)', category: 'cranial', color: '#EF4444' },
  // 2: B (Supramental) - Mandibular
  { id: 2, name: 'B-point (Supramental)', category: 'mandibular', color: '#3B82F6' },
  // 3: Me (Menton) - Mandibular
  { id: 3, name: 'Menton (Me)', category: 'mandibular', color: '#3B82F6' },
  // 4: N (Nasion) - Cranial
  { id: 4, name: 'Nasion (N)', category: 'cranial', color: '#EF4444' },
  // 5: Or (Orbitale) - Cranial
  { id: 5, name: 'Orbitale (Or)', category: 'cranial', color: '#EF4444' },
  // 6: Pog (Pogonion óseo) - Mandibular
  { id: 6, name: 'Pogonion (Pog)', category: 'mandibular', color: '#3B82F6' },
  // 7: PNS (Posterior Nasal Spine) - Cranial
  { id: 7, name: 'PNS (Posterior Nasal Spine)', category: 'cranial', color: '#EF4444' },
  // 8: Pn (Pronasale / Nose Tip) - Soft Tissue
  { id: 8, name: 'Pronasale (Pn)', category: 'soft-tissue', color: '#EAB308' },
  // 9: R (Ridge) - Mandibular
  { id: 9, name: 'Ridge (R)', category: 'mandibular', color: '#3B82F6' },
  // 10: S (Sella) - Cranial
  { id: 10, name: 'Sella (S)', category: 'cranial', color: '#EF4444' },
  // 11: Ar (Articulare) - Cranial
  { id: 11, name: 'Articular (Ar)', category: 'cranial', color: '#EF4444' },
  // 12: Co (Condylion) - Mandibular
  { id: 12, name: 'Condylion (Co)', category: 'mandibular', color: '#3B82F6' },
  // 13: Go (Gonion) - Mandibular
  { id: 13, name: 'Gonion (Go)', category: 'mandibular', color: '#3B82F6' },
  // 14: Po (Porion) - Cranial
  { id: 14, name: 'Porion (Po)', category: 'cranial', color: '#EF4444' },
  // 15: LPM (Low Point Mandible) - Mandibular
  { id: 15, name: 'Low Point Mandible (LPM)', category: 'mandibular', color: '#3B82F6' },
  // 16: LMT (Lower Molar Tip) - Mandibular
  { id: 16, name: 'Lower Molar Tip (LMT)', category: 'mandibular', color: '#3B82F6' },
  // 17: UMT (Upper Molar Tip) - Dental
  { id: 17, name: 'Upper Molar Tip (UMT)', category: 'dental', color: '#22C55E' },
  // 18: UPM (Upper 2nd PM Cusp) - Dental
  { id: 18, name: 'Upper 2nd PM Cusp (UPM)', category: 'dental', color: '#22C55E' },
  // 19: UIA (Upper Incisor Apex) - Dental
  { id: 19, name: 'Upper Incisor Apex (UIA)', category: 'dental', color: '#22C55E' },
  // 20: UIT (Upper Incisor Tip) - Dental
  { id: 20, name: 'Upper Incisor Tip (UIT)', category: 'dental', color: '#22C55E' },
  // 21: UPM2 (Upper 2nd PM Cusp 2) - Dental (DUPLICADO en Aariz)
  { id: 21, name: 'Upper 2nd PM Cusp 2 (UPM2)', category: 'dental', color: '#22C55E' },
  // 22: LIA (Lower Incisor Apex) - Dental
  { id: 22, name: 'Lower Incisor Apex (LIA)', category: 'dental', color: '#22C55E' },
  // 23: LIT (Lower Incisor Tip) - Dental
  { id: 23, name: 'Lower Incisor Tip (LIT)', category: 'dental', color: '#22C55E' },
  // 24: LPM2 (Lower 2nd PM Cusp) - Dental (DUPLICADO en Aariz)
  { id: 24, name: 'Lower 2nd PM Cusp (LPM2)', category: 'dental', color: '#22C55E' },
  // 25: Ls (Labrale Superius) - Soft Tissue
  { id: 25, name: 'Labrale Superius (Ls)', category: 'soft-tissue', color: '#EAB308' },
  // 26: Sn (Subnasale) - Soft Tissue
  { id: 26, name: 'Subnasale (Sn)', category: 'soft-tissue', color: '#EAB308' },
  // 27: Pog' (Soft Pogonion) - Soft Tissue
  { id: 27, name: 'Soft Pogonion (Pog\')', category: 'soft-tissue', color: '#EAB308' },
  // 28: Glabela - Soft Tissue
  { id: 28, name: 'Glabela', category: 'soft-tissue', color: '#EAB308' },
]

export const CRANIAL_LANDMARKS = LANDMARKS.filter(l => l.category === 'cranial')
export const MANDIBULAR_LANDMARKS = LANDMARKS.filter(l => l.category === 'mandibular')
export const DENTAL_LANDMARKS = LANDMARKS.filter(l => l.category === 'dental')
export const SOFT_TISSUE_LANDMARKS = LANDMARKS.filter(l => l.category === 'soft-tissue')

export default LANDMARKS
